docs: Add a provider framework guide.

paddycarver opened the pull request:
Document how to write a new provider for Terraform, or contribute new
pieces or functionality to an existing provider.

This doc should be largely feature complete. Things I would have liked
to included that didn't make it into this cut:

* How to make resources importable
* How to test resources, data sources, migrations, importing, etc.
* A better explanation of the configuration of a provider, separating
  out the Schema property and the ConfigureFunc more clearly.
* A better diagram on the lifecycle page.
* How to contribute documentation for your provider to the website.
* Useful practices for log messages.
* Notes on error wrapping
* Using timeouts
* Where, _exactly_, Exists fits into the lifecycle and when it should be used

While these things would be great to have, I'm calling this the MVP,
with an intent to add that information over time (or other people can PR
it in). For now, I think what we have is valuable enough to get it
published.

Because of the size of this PR, and the number of people involved in reviewing
it, it'd be great if reviews could be restricted to just the things that should
block a merge. Suggestions for enhancements or future directions are welcome,
but it'd be great to keep those as comments. Just keep in mind that changes
require people to re-review, that's all. :)

sethvargo commented on website/source/layouts/docs.erb:
```diff
@@ -552,6 +552,38 @@
+            <a href="/docs/internals/providers/index.html">Writing Your Own Provider</a>
+            <ul class="nav">
+              <li<%= sidebar_current("docs-internals-provider-guide-lifecycle") %>>
+                <a href="/docs/internals/providers/lifecycle.html">Understanding the Terraform Lifecycle</a>
```
This is really long for the sidebar. I'd say "Terraform Lifecycle"

sethvargo commented on website/source/layouts/docs.erb:
```diff
@@ -552,6 +552,38 @@
+                <a href="/docs/internals/providers/lifecycle.html">Understanding the Terraform Lifecycle</a>
+              </li>
+              <li<%= sidebar_current("docs-internals-provider-guide-vs-resources") %>>
+                <a href="/docs/internals/providers/providers-v-resources.html">Providers vs. Resources & Data Sources</a>
```
Also, this is a longish title. Can we shorten?

paddycarver replied:
I couldn't find a way to shorten this, and looking at the section, it wasn't really... saying much. I've just deleted it and the link in the sidebar.

sethvargo commented on website/source/layouts/docs.erb:
```diff
@@ -552,6 +552,38 @@
+                <a href="/docs/internals/providers/providers-v-resources.html">Providers vs. Resources & Data Sources</a>
+              </li>
+              <li<%= sidebar_current("docs-internals-provider-guide-new-provider") %>>
+                <a href="/docs/internals/providers/new-provider.html">Creating a New Provider</a>
```
I think these can be "Creating Providers" and "Creating Resources" for brevity.

paddycarver commented on website/source/layouts/docs.erb:
```diff
@@ -552,6 +552,38 @@
           <li<%= sidebar_current("docs-internals-plugins") %>>
             <a href="/docs/internals/internal-plugins.html">Internal Plugins</a>
           </li>
+					<li<%= sidebar_current("docs-internals-provider-guide") %>>
```
Hm, weird, thought I caught all those. Thanks! I'll fix that.

paddycarver commented:
Thanks for the speedy review @sethvargo. I'm pretty sure I addressed your concerns, and tried to apply them consistently.

sethvargo commented on website/source/docs/internals/providers/new-data-source.html.md:
```diff
@@ -0,0 +1,90 @@
+
+# Creating Data Sources
+
+Once we have a Provider, we can add some Data Sources to that provider, which
```
There's a lot of "we" and "you" here. You are writing technical guides, so this is generally frowned upon.

sethvargo commented on website/source/docs/internals/providers/new-data-source.html.md:
```diff
@@ -0,0 +1,90 @@
+Provider, and this kind of naming scheme really helps keep things organised and
+easy to find.
+
+## Registering the Data Source  With the Provider
```
There's an extra space after "Souce" and "With" should not be capitalized

sethvargo commented on website/source/docs/internals/providers/new-data-source.html.md:
```diff
@@ -0,0 +1,90 @@
+If you want to contribute your code back to the Terraform repo, it makes things
+easier if you follow this convention.
+
+Inside this file, you’re going to define a function that takes no arguments and
```
Do we have code samples for this? It's really challenging to follow along the walls of text without some copy-paste examples or sample code.

paddycarver replied:
I don't have code samples at the moment, and do not believe I'd be able to come up with samples of the appropriate quality to meet the business timeline for publishing this. (~sometime next week, earlier the better).

sethvargo commented on website/source/docs/internals/providers/new-data-source.html.md:
```diff
@@ -0,0 +1,90 @@
+
+## Defining the Data Source Properties
+
+Each Data Source has a set of properties called the “schema” that is stored in
```
Fancy quotes. Fancy quotes. Fancy quotes. Please do a mass find-and-replace. These will make the builder sad and is a by-product of working in a Google doc. You can disable fancy quotes in your google settings.

sethvargo commented on website/source/docs/internals/providers/new-data-source.html.md:
```diff
@@ -0,0 +1,90 @@
+## Registering the Data Source  With the Provider
+
+Now that you’ve got your function, switch back to your Provider definition, and
+add the Data Source to the Provider’s `DataSourcesMap` property. The key will
```
I find it weird that we keep capitalizing "Data Sources" and "Resources"

sethvargo commented on website/source/docs/internals/providers/new-data-source.html.md:
```diff
@@ -0,0 +1,90 @@
+behaviour](/docs/internals/providers/schema.html) for resources that helps
+Terraform do the right thing without you needing to write a bunch of code.
+
+~> **Note:** “id” is a reserved property name. Don’t call your property “id”.
```
Another fancy quote, this time it's the single quote in "Don't".

sethvargo commented on website/source/docs/internals/providers/new-data-source.html.md:
```diff
@@ -0,0 +1,90 @@
+`ConfigureFunc`. Typically, this is where you’d put the configured API client,
+for example.
+
+The `Read` function should then use the passed `*ResourceData` to retrieve the
```
PV

sethvargo commented on website/source/docs/internals/providers/new-provider.html.md:
```diff
@@ -0,0 +1,89 @@
+
+# Creating Providers
+
+Say you’ve found a new API provider that Terraform doesn’t support, and you
```
This isn't really in line with the technical documentation we author.

sethvargo commented on website/source/docs/internals/providers/new-provider.html.md:
```diff
@@ -0,0 +1,89 @@
+contributing this to the Terraform codebase, create a directory named after
+your plugin in the `builtin/providers` directory of Terraform. If you’re
+planning on maintaining it outside of Terraform as a standalone plugin, create
+a new repository for it-we recommend
```
You'll still want spaces around these dashes - the markdown parser will convert them

sethvargo commented on website/source/docs/internals/providers/new-provider.html.md:
```diff
@@ -0,0 +1,89 @@
+## Registering Your Provider
+
+Because Terraform uses a plugin architecture, we need to register our new
+plugin with Terraform before Terraform knows how to use it. (Fun fact: even the
```
Remove the fun fact

sethvargo commented on website/source/docs/internals/providers/new-resource.html.md:
```diff
@@ -0,0 +1,128 @@
+
+## Registering the Resource With the Provider
+
+Now that you’ve got your function, switch back to your Provider definition, and
```
As I said above, I don't think type of language is appropriate for technical documentation or guides.

sethvargo replied:
Provider probably shouldn't be capitalized either.

paddycarver commented:
Thanks for the in-depth feedback, @sethvargo, really appreciate it. I've addressed most your points, I believe--hyphens should now have spaces, magic quotes have been eliminated, paragraphs have become lists, images have been added.

There are a lot of voice concerns here and they're all valid and I want to address them all. However, because this is a writing voice I don't have much practice in, I've updated the following files to try and make the changes you've requested. Could you just take a look at those (ignoring the voice in the others for the moment) and let me know if you're happy with those changes? If so, I'll update the rest. I just want to be sure I've got this right before I spend more time applying it.

The files with updated voice:

* async.html
* index.html
* lifecycle.html
* new-data-source.html

sethvargo commented on website/source/docs/internals/providers/index.html.md:
```diff
@@ -0,0 +1,30 @@
+One way to think of the separation is that the core tells the providers what to
+do, and the providers make that change against the APIs.
+
+To hook into this behaviour, Terraform has a provider framework. This guide
```
This is the "british" spelling of "behavior". I think we should be consistent and use the "american" one instead.

sethvargo commented on website/source/docs/internals/providers/lifecycle.html.md:
```diff
@@ -0,0 +1,38 @@
+
+# Terraform Lifecycle
+
+Terraform runs follow a predictable lifecycle:
```
What is a run?

(I know, but it'd be helpful to provide a bit more context here, since we usually say "plan" or "apply" exlicitly)

sethvargo commented on website/source/docs/internals/providers/new-data-source.html.md:
```diff
@@ -0,0 +1,86 @@
+
+# Creating Data Sources
+
+Data sources are components of providers that pull information from outside of
```
This isn't necessarily true. Data sources pull data from external sources, but those externals sources could be managed by Terraform too. This makes it sound mutually exclusive.

sethvargo commented on website/source/docs/internals/providers/new-data-source.html.md:
```diff
@@ -0,0 +1,86 @@
+# Creating Data Sources
+
+Data sources are components of providers that pull information from outside of
+Terraform's managed resources, to make it available in configuration.  They
```
There is a double space after period here.

sethvargo commented on website/source/docs/internals/providers/new-data-source.html.md:
```diff
@@ -0,0 +1,86 @@
+Terraform's managed resources, to make it available in configuration.  They
+should exist in the same package as the provider.
+
+The built-in plugins for Terraform that are shipped as part of the codebase
```
This reads choppy. I would remove "that are shipped as part of the codebase" or move it to its own sentence.

sethvargo commented on website/source/docs/internals/providers/new-data-source.html.md:
```diff
@@ -0,0 +1,86 @@
+
+The built-in plugins for Terraform that are shipped as part of the codebase
+have a file naming guideline: they use
+`data_source_{api}_{data_source_name}.go` as the template to name files, and
```
We use the convention `<THING>` in these cases. So `data_source_<API>_<DATA_SOURCE_NAME>` instead

sethvargo commented on website/source/docs/internals/providers/new-data-source.html.md:
```diff
@@ -0,0 +1,86 @@
+easier to contribute code back to the Terraform codebase.
+
+Just like resources, data sources are functions that take no arguments and
+return a `*schema.Resource`. While functions can technically be named anything,
```
Remove "technically"

sethvargo commented on website/source/docs/internals/providers/new-data-source.html.md:
```diff
@@ -0,0 +1,86 @@
+`*schema.Provider`'s `ConfigureFunc`. It generally contains a configured API
+client or similar form of access for the provider.
+
+The `Read` function should use the passed `*ResourceData` to retrieve the ID
```
should use -> uses

sethvargo commented on website/source/docs/internals/providers/new-provider.html.md:
```diff
@@ -0,0 +1,89 @@
+
+# Creating Providers
+
+Say you've found a new API provider that Terraform doesn't support, and you
```
There's still a lot of first person in this entire page.

paddycarver replied:
Yeah, I mentioned I had only updated the first few pages. Sorry for the confusion, will try to be clearer next time. Was trying to save us both some time and spot-check for the voice you wanted before I rewrote a 7000 word document (or made you re-edit a 7000 word document).

sethvargo commented on website/source/docs/internals/providers/new-resource.html.md:
```diff
@@ -0,0 +1,127 @@
+code in the package your Provider is defined in.
+
+In the built-in plugins for Terraform that are shipped as part of the codebase,
+we have a file naming guideline: we use `resource_{resource_name}.go` as the
```
`<THING>`

sethvargo commented on website/source/docs/internals/providers/new-resource.html.md:
```diff
@@ -0,0 +1,127 @@
+
+## Registering the Resource With the Provider
+
+Now that you've got your function, switch back to your Provider definition, and
```
"You've got your"

sethvargo commented on website/source/docs/internals/providers/new-resource.html.md:
```diff
@@ -0,0 +1,127 @@
+add the Resource to the Provider's `ResourcesMap` property. The key will be the
+name of the Resource used in state and configuration files, and (by convention)
+should match the `{resource_name}` part used in the filename for the file that
+defines the Resource. Nothing will necessarily break if you don't do this, but
```
"Nothing will... if you don't" - double negative and challenging to follow

sethvargo commented on website/source/docs/internals/providers/new-resource.html.md:
```diff
@@ -0,0 +1,127 @@
+name of the Resource used in state and configuration files, and (by convention)
+should match the `{resource_name}` part used in the filename for the file that
+defines the Resource. Nothing will necessarily break if you don't do this, but
+if you happen to accidentally have a collision with any other resource in any
```
happen to accidentally -> accidentally

sethvargo replied:
have a collision -> collide

sethvargo commented on website/source/docs/internals/providers/new-resource.html.md:
```diff
@@ -0,0 +1,127 @@
+
+## Calling the API
+
+Now that we have a Resource fully defined, it's time to make it do something.
```
Now that the resource is fully defined, it's time to add functionality to manage upstream resources.

sethvargo commented on website/source/docs/internals/providers/new-resource.html.md:
```diff
@@ -0,0 +1,127 @@
+`ConfigureFunc`. Typically, this is where you'd put the configured API client,
+for example.
+
+Of all these functions, only the `Update` and `Exists` functions are optional.
```
There are definitely other functions on the helper/schema - `Importer`, `Timeouts`, `MigrateState`

paddycarver replied:
I'm not sure what this is suggesting. `Timeout` and `MigrateState` do not call the API, so I'm not sure they would belong under a section titled "Calling the API". `Importer` _could_ be considered part of this, but would likely make more sense in its own section, given its lifecycle is a bit... different. `MigrateState` is talked about in the `schema.html` page, and `Importer` and `Timeouts` were both punted on for this revision.

sethvargo commented on website/source/docs/internals/providers/new-resource.html.md:
```diff
@@ -0,0 +1,127 @@
+that cannot be updated. If the `Exists` function isn't set, Terraform will not
+check if the Resource already/still exists.
+
+The `Create` function should use the passed `*ResourceData` to make an API call
```
Reading this (as someone who has written a lot of Terraform), I'm still not really sure what I do with this information. I think adding code examples and method signatures here is very important.

sethvargo commented on website/source/docs/internals/providers/resource-data.html.md:
```diff
@@ -0,0 +1,180 @@
+
+## How to Conceptualise `ResourceData`
+
+The most useful way to think of `ResourceData` is not as a placeholder for any
```
I read this sentence 3 times and I'm still not sure what is being said. Could you rephrase or break into separate sentences please?

sethvargo commented on website/source/docs/internals/providers/resource-data.html.md:
```diff
@@ -0,0 +1,180 @@
+## How to Conceptualise `ResourceData`
+
+The most useful way to think of `ResourceData` is not as a placeholder for any
+one concept in Terraform—it's not just information about the config or the
```
Space around `-`

sethvargo commented on website/source/docs/internals/providers/resource-data.html.md:
```diff
@@ -0,0 +1,180 @@
+* The diff being applied
+* Any calls to `ResourceData.Set`
+
+Those are expressed in the order of their priority. That is, the config
```
Remove `that is`

sethvargo commented on website/source/docs/internals/providers/resource-data.html.md:
```diff
@@ -0,0 +1,180 @@
+
+## Retrieving Properties
+
+To retrieve properties from `ResourceData`, use the `Get` or `GetOk` methods.
```
Again, I see lots of text, but what is `GetOk`? How do I use it? Adding a 2-3 line code sample here would do wonders for the usefulness of this guide.

sethvargo commented on website/source/docs/internals/providers/resource-data.html.md:
```diff
@@ -0,0 +1,180 @@
+your schema, but not in the config, `Get` returns the empty value for that
+type.
+
+`ResourceData` also has a `GetOk` method that functions identically to its
```
if it functioned identically, it would return the same thing. It functions _simarily_

sethvargo commented on website/source/docs/internals/providers/resource-data.html.md:
```diff
@@ -0,0 +1,180 @@
+type.
+
+`ResourceData` also has a `GetOk` method that functions identically to its
+`Get` method, but with an extra return parameter. The new return parameter
```
s/new/extra

sethvargo commented on website/source/docs/internals/providers/resource-data.html.md:
```diff
@@ -0,0 +1,180 @@
+something modifying the infrastructure outside of Terraform) or from config
+changes. The `GetChange` method returns two values; a basic understanding is
+that the first value is what the property _used to be_, and the second value is
+what the property _should be changed to_.  A slightly more nuanced
```
There are two spaces after this period

sethvargo commented on website/source/docs/internals/providers/resource-data.html.md:
```diff
@@ -0,0 +1,180 @@
+method is provided. It takes the key of a property, just like `Get`, as an
+argument, and the value to set it to.
+
+It's important that all your properties get set using the `Set` method, as
```
"get set using the `Set` method". Please try to rephrase this using words that aren't also keywords in this context.

sethvargo commented on website/source/docs/internals/providers/resource-data.html.md:
```diff
@@ -0,0 +1,180 @@
+
+## Partial Updates
+
+Some resources and data sources cannot be managed using only a single API call;
```
This isn't a semicolon kind of thing. These are two separate thoughts.

sethvargo commented on website/source/docs/internals/providers/resource-data.html.md:
```diff
@@ -0,0 +1,180 @@
+## Partial Updates
+
+Some resources and data sources cannot be managed using only a single API call;
+for example, a resource may need its permissions set or a compute instance may
```
its' permissions (permission belongs to the resource)

sethvargo commented on website/source/docs/internals/providers/resource-data.html.md:
```diff
@@ -0,0 +1,180 @@
+
+Some resources and data sources cannot be managed using only a single API call;
+for example, a resource may need its permissions set or a compute instance may
+need a disk attached to it, even though the instance and disk are modeled as a
```
drop "to it"

sethvargo commented on website/source/docs/internals/providers/resource-data.html.md:
```diff
@@ -0,0 +1,180 @@
+Should the function return an error, the `*ResourceData` gets discarded and is
+_not_ set as the state.
+
+The `Partial` method accepts a boolean designating whether partial state mode
```
Again, I really do not think this guide is useful without code samples. You're basically describing the method signature in a wall of text that could easily be replaced by a code sample of 5-6 lines with a comment.

sethvargo commented on website/source/docs/internals/providers/resource-data.html.md:
```diff
@@ -0,0 +1,180 @@
+the `SetId` method, passing the string to use as an ID. Only strings may be
+used as IDs.
+
+You should always set the ID as soon as you possibly can. Even with
```
"as soon as you possibly can" -> as soon as possible

sethvargo commented on website/source/docs/internals/providers/resources-vs-data-sources.html.md:
```diff
@@ -0,0 +1,21 @@
+
+# Resources vs Data Sources
+
+Resources and Data Sources are incredibly similar, and for the most part, can
```
s/incredibly/very

sethvargo commented on website/source/docs/internals/providers/schema.html.md:
```diff
@@ -0,0 +1,322 @@
+
+There are some fields on an Resource or Data Source that can't be influenced by
+the user at all, which the API provider is the source of truth for. Things like
+the timestamp a resource was created, or the ID of a resource. For these
```
This isn't a sentence

paddycarver commented:
Closing in favour if #14250.
